### PR TITLE
Cypress automation

### PR DESCRIPTION
**Pull request by agorthi-akamai**
## Description 📝
Highlight the Pull Request's context and intentions.

## Changes  🔄
List any change relevant to the reviewer.
- ...
- ...

## Target release date 🗓️
Please specify a release date to guarantee timely review of this PR. If exact date is not known, please approximate and update it as needed.

## Preview 📷
**Include a screenshot or screen recording of the change**

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
| 📷 | 📷 |

### Prerequisites
(How to setup test environment)
- ...
- ...

### Reproduction steps
(How to reproduce the issue, if applicable)
- ...
- ...

### Verification steps
(How to verify changes)
- ...
- ...

## As an Author I have considered 🤔

*Check all that apply*

- [ ] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [ ] 🤏 Splitting feature into small PRs
- [ ] ➕ Adding a changeset
- [ ] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

---
## Commit message and pull request title format standards

> **Note**: Remove this section before opening the pull request
**Make sure your PR title and commit message on squash and merge are as shown below**

`<commit type>: [JIRA-ticket-number] - <description>`

**Commit Types:**
- `feat`: New feature for the user (not a part of the code, or ci, ...).
- `fix`: Bugfix for the user (not a fix to build something, ...).
- `change`: Modifying an existing visual UI instance. Such as a component or a feature.
- `refactor`: Restructuring existing code without changing its external behavior or visual UI. Typically to improve readability, maintainability, and performance.
- `test`: New tests or changes to existing tests. Does not change the production code.
- `upcoming`: A new feature that is in progress, not visible to users yet, and usually behind a feature flag.

**Example:** `feat: [M3-1234] - Allow user to view their login history`

---

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -23,8 +23,8 @@ lib
 restart.sh
 
 # dotenv environment variables file
-.env*
```
Comment (venkymano-akamai) [inline .gitignore:26]: don't commit git ignore

File: packages/manager/.env [hunk 2]
```diff
@@ -0,0 +1,67 @@
+## UNCOMMENT & UPDATE THESE ENVIRONMENT VARIABLES WITH YOUR INFO,
+## THEN SAVE IN packages/manager AS `.env`
+
+##################################
+# Required to run the app:
+##################################
+
+REACT_APP_LOGIN_ROOT='https://login.linode.com'
+REACT_APP_API_ROOT='https://api.linode.com/v4'
```
Comment (venkymano-akamai) [inline packages/manager/.env:9]: don't commit .env

File: packages/manager/cypress/e2e/core/cloudpulse/widget-verification.spec.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    validateWidgetTitle,
+    setGranularity,
+    setAggregation,
+    verifyGranularity,
+    verifyAggregation,
+    verifyZoomInOut,
+    applyGlobalRefresh,
+
+  } from 'support/util/cloudpulse';
+  import { widgets } from 'support/constants/widgets';
```
Comment (venkymano-akamai) [inline packages/manager/cypress/e2e/core/cloudpulse/widget-verification.spec.ts:15]: do you have installed eslint plugin and formatted these

File: packages/manager/cypress/e2e/core/linodes/clone-linode.spec.ts [hunk 4]
```diff
@@ -46,6 +46,7 @@ describe('clone linode', () => {
    * - Confirms Linode Clone flow via the Linode details page.
    * - Confirms that Linode can be cloned successfully.
    */
+
```
Comment (venkymano-akamai) [inline packages/manager/cypress/e2e/core/linodes/clone-linode.spec.ts:49]: Just remove this change, no changes in this file

File: packages/manager/cypress/support/ui/autocomplete.ts [hunk 5]
```diff
@@ -13,6 +13,26 @@ export const autocomplete = {
...
+  findByPlaceholderCustom: (
+    title: string,
+    options?: SelectorMatcherOptions
+  ): Cypress.Chainable => {
+    return cy
+    .get("[data-qa-autocomplete] input[placeholder='" + title + "']");
+  },
+
+  /**
+   * Finds a autocomplete that has the given placehoder text.
```
Comment (venkymano-akamai) [inline packages/manager/cypress/support/ui/autocomplete.ts:28]: Finds a autocomplete that has the given data-qa-autocomplete value

File: packages/manager/cypress/support/ui/autocomplete.ts [hunk 6]
```diff
@@ -13,6 +13,26 @@ export const autocomplete = {
   find: (): Cypress.Chainable => {
     return cy.get('[data-qa-autocomplete] input');
   },
+  /**
+   * Finds a autocomplete that has the given placehoder text.
```
Comment (venkymano-akamai) [inline packages/manager/cypress/support/ui/autocomplete.ts:17]: Finds a input in the autocomplete that has the given placehoder text.

File: packages/manager/cypress/support/ui/buttons.ts [hunk 7]
```diff
@@ -47,6 +47,14 @@ export const buttonGroup = {
   find: (): Cypress.Chainable => {
     return cy.get('[data-qa-buttons="true"]');
   },
+  /**
+   * Finds a refresh button  and returns the Cypress chainable.
+   *
+   * @returns Cypress chainable.
+   */
+  findRefreshIcon: (): Cypress.Chainable => {
+    return cy.get('[data-qa-refresh-button="true"]');
```
Comment (venkymano-akamai) [inline packages/manager/cypress/support/ui/buttons.ts:56]: We don't need to keep this in button, rather we create a cloudpulse.ts and keep it there

File: packages/manager/cypress/support/ui/buttons.ts [hunk 8]
```diff
@@ -61,4 +69,14 @@ export const buttonGroup = {
...
       .closest('button');
   },
+  /**
+   * Finds a button within a button group by its title and returns the Cypress chainable.
+   *
+   * @param buttonTitle - Title of button to find.
+   *
+   * @returns string.
+   */
+  findZoomButtonByTitle: (buttonTitle: string): string => {
```
Comment (venkymano-akamai) [inline packages/manager/cypress/support/ui/buttons.ts:79]: Same as above

File: packages/manager/cypress/e2e/core/cloudpulse/dashboard-filters.spec.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  beforeEach(() => {
+    navigateToCloudpulse();
+   });
+
+it('should verify cloudpulse availability when feature flag is set to false', () => {
+    visitCloudPulseWithFeatureFlagsDisabled();
+});
+
+it('should clear the preferences of the dashboard', () => {
+  verifyZerothPage(dashboardName);
```
Comment (venkymano-akamai) [inline packages/manager/cypress/e2e/core/cloudpulse/dashboard-filters.spec.ts:38]: Did we check how to run this only once before all our test cases

File: packages/manager/cypress/support/constants/aggregation.ts [hunk 10]
```diff
@@ -0,0 +1,11 @@
+export enum aggregation {
+  Avg = 'avg',
+  Max = 'max',
+  Min = 'min',
+  Sum = 'sum',
+}
+
+export const aggregationConfig = {
+  all: [aggregation.Avg, aggregation.Max, aggregation.Min, aggregation.Sum],
```
Comment (venkymano-akamai) [inline packages/manager/cypress/support/constants/aggregation.ts:9]: is it possible to make this readonly

File: packages/manager/cypress/support/constants/aggregation.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+export enum aggregation {
+  Avg = 'avg',
+  Max = 'max',
+  Min = 'min',
+  Sum = 'sum',
+}
+
+export const aggregationConfig = {
```
Comment (venkymano-akamai) [inline packages/manager/cypress/support/constants/aggregation.ts:8]: Also, here we can use snake_case like AGGREGATION_CONFIG

File: packages/manager/cypress/support/ui/buttons.ts [hunk 12]
```diff
@@ -47,7 +47,6 @@ export const buttonGroup = {
   find: (): Cypress.Chainable => {
     return cy.get('[data-qa-buttons="true"]');
   },
-
```
Comment (venkymano-akamai) [inline packages/manager/cypress/support/ui/buttons.ts:50]: Why this file is coming, it should not come in the PR

File: packages/manager/cypress/support/util/cloudpulse.ts [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+};
+
+/**
+ * Selects a resource name from the resources dropdown and verifies the selection.
+ * @param {string} service - The name of the service to select.
+ */
+export const selectAndVerifyServiceName = (service: string) => {
+  const resourceInput = ui.autocomplete.findByTitleCustom('Select Resources');
+  resourceInput.findByTitle('Open').click();
+  resourceInput.click().type(`${service}{enter}`);
```
Comment (venkymano-akamai) [inline packages/manager/cypress/support/util/cloudpulse.ts:154]: FYI, this is not working with autocompletepopper methods

File: packages/manager/cypress/support/util/cloudpulse.ts [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+  cy.get(widgetSelector)
+    .first()
+    .scrollIntoView()
+    .should('be.visible')
+    .within(() => {
+      ui.autocomplete
+        .findByTitleCustom('Select an Interval')
+        .findByTitle('Open')
+        .click();
+      ui.autocompletePopper
```
Comment (venkymano-akamai) [inline packages/manager/cypress/support/util/cloudpulse.ts:268]: Can't we just use, existing  ui.autocompletePopper method here

File: packages/manager/cypress/support/util/cloudpulse.ts [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+  cy.get(widgetSelector)
+    .first()
+    .scrollIntoView()
+    .should('be.visible')
+    .within(() => {
+      ui.autocomplete
+        .findByTitleCustom('Select an Aggregate Function')
+        .findByTitle('Open')
+        .click();
+      ui.autocompletePopper
```
Comment (venkymano-akamai) [inline packages/manager/cypress/support/util/cloudpulse.ts:308]: Same as above

File: .gitignore (hunk 1 above)
Comment (agorthi-akamai) [reply .gitignore]: fixed

File: packages/manager/.env (hunk 2 above)
Comment (agorthi-akamai) [reply packages/manager/.env]: fixed

File: packages/manager/cypress/e2e/core/cloudpulse/widget-verification.spec.ts (hunk 3 above)
Comment (agorthi-akamai) [reply packages/manager/cypress/e2e/core/cloudpulse/widget-verification.spec.ts]: fixed and formated the code by eslint plugin

File: packages/manager/cypress/e2e/core/linodes/clone-linode.spec.ts (hunk 4 above)
Comment (agorthi-akamai) [reply packages/manager/cypress/e2e/core/linodes/clone-linode.spec.ts]: remove the extra line

File: packages/manager/cypress/support/ui/buttons.ts (hunk 7 above)
Comment (agorthi-akamai) [reply packages/manager/cypress/support/ui/buttons.ts]: Implement CloudPulse class in new cloudpulse.ts file under ui folder

File: packages/manager/cypress/e2e/core/cloudpulse/dashboard-filters.spec.ts (hunk 9 above)
Comment (agorthi-akamai) [reply packages/manager/cypress/e2e/core/cloudpulse/dashboard-filters.spec.ts:38]: I've looked at every alternative, and this appears to be fine.

File: packages/manager/cypress/support/constants/aggregation.ts (hunk 11 above)
Comment (agorthi-akamai) [reply packages/manager/cypress/support/constants/aggregation.ts:8]: ignored this comment as you had suggested.

File: packages/manager/cypress/support/ui/buttons.ts (hunk 12 above)
Comment (agorthi-akamai) [reply packages/manager/cypress/support/ui/buttons.ts]: fixed

File: packages/manager/cypress/support/util/cloudpulse.ts (hunk 13 above)
Comment (agorthi-akamai) [reply packages/manager/cypress/support/util/cloudpulse.ts:155]: know issue

File: packages/manager/cypress/support/util/cloudpulse.ts (hunk 14 above)
Comment (agorthi-akamai) [reply packages/manager/cypress/support/util/cloudpulse.ts]: I've implemented a new method in consideration of your suggestions.

File: packages/manager/cypress/support/util/cloudpulse.ts (hunk 15 above)
Comment (agorthi-akamai) [reply packages/manager/cypress/support/util/cloudpulse.ts]: I've implemented a new method in consideration of your suggestions.

File: packages/manager/cypress/support/ui/buttons.ts (hunk 12 above)
Comment (agorthi-akamai) [reply packages/manager/cypress/support/ui/buttons.ts]: "Resolved the issue by removing buttons.ts from the current commit

File: .gitignore (hunk 1 above)
Comment (agorthi-akamai) [reply .gitignore]: "Resolved the issue by removing "git ignore" from the current commit

File: packages/manager/.env (hunk 2 above)
Comment (agorthi-akamai) [reply packages/manager/.env]: "Resolved the issue by removing ".env" from the current commit

**Pull request closed without merging**